spec/unit/matrix-client.spec.ts

```diff
@@ -231,6 +287,130 @@ describe("MatrixClient", function() {
         client.stopClient();
     });
 
+    describe('timestampToEvent', () => {
+        const roomId = '!room:server.org';
+        const eventId = "$eventId:example.org";
+        const unstableMsc3030Prefix = "/_matrix/client/unstable/org.matrix.msc3030";
+
+        it('should call stable endpoint', async () => {
+            httpLookups = [{
+                method: "GET",
+                path: `/rooms/${encodeURIComponent(roomId)}/timestamp_to_event`,
+                data: { event_id: eventId },
+                expectQueryParams: {
+                    ts: '0',
+                    dir: 'f',
+                },
+            }];
+
+            await client.timestampToEvent(roomId, 0, 'f');
+
+            expect(client.http.authedRequest.mock.calls.length).toStrictEqual(1);
+            const [method, path, queryParams,, { prefix }] = client.http.authedRequest.mock.calls[0];
+            expect(method).toStrictEqual('GET');
+            expect(prefix).toStrictEqual(ClientPrefix.V1);
+            expect(path).toStrictEqual(
+                `/rooms/${encodeURIComponent(roomId)}/timestamp_to_event`,
+            );
+            expect(queryParams).toStrictEqual({
+                ts: '0',
+                dir: 'f',
+            });
+        });
+
+        it('should fallback to unstable endpoint when no support for stable endpoint', async () => {
+            httpLookups = [{
+                method: "GET",
+                path: `/rooms/${encodeURIComponent(roomId)}/timestamp_to_event`,
+                prefix: ClientPrefix.V1,
+                error: {
+                    httpStatus: 404,
```

Comment:
we should also be testing `405 M_UNRECOGNIZED` per https://github.com/matrix-org/matrix-spec-proposals/pull/3743

Reply:
I don't see the benefit to this test. We use the same `GET` method for stable and unstable.

Reply:
The code says we're following MSC3743 though? We should be making sure we actually are...

Reply:
That MSC isn't set up to be broken down like that though - we really do need a test case here, sorry.

Reply:
This isn't something I will be addressing with the current reasoning.

Reply:
The current reasoning is spec/MSC compliance, which is a blocker for this PR landing, sorry.

Reply:
Fixed in 628bcbf33a3dfa3d685ee5bf4bc933beca30590f

Reply:
Adding support for `405` seems like the wrong move. If we ever see a `405` response here, then this is most likely an issue with Element itself (we used the wrong method in our first API call against the stable endpoint) and we should try to fallback. It was better to blow-up the app so the user has feedback to report or ideally auto-report the issue.

Reply:
@andybalaam came up with a great example for a slightly different case but makes this code make more sense. Imagine a scenario where the server supports `GET /foo` and there is a MSC that went stable for `POST /foo`. In this case, making a `POST /foo` could return a `405` unsupported method error and we would want to fallback to the unstable endpoint.

In the `/timestamp_to_event` case because only one HTTP method is supported, it's more obvious that a `405` would be an error in our code, but we can opt to leave the general pattern here ⏩